File: cpp/enigma.cpp
#include "enigma.h"
#include "utils.h"
Enigma::Enigma(const std::string& rotorspec, const std::string& ring_settings, const std::string& plugboard_settings) 
: plugboard(plugboard_settings)
{
    std::vector<std::string> rotor_names = split(rotorspec);
    for (int i = 0; i < 3; i++)
    {
        rotors.push_back(RotorFactory(rotor_names[i], ring_settings[i]));
    }
    reflector = WheelFactory("UKW-B");
}
std::string Enigma::get_window_positions() const
{
    std::string results;
    results += rotors[0].get_window_position();
    results += rotors[1].get_window_position();
    results += rotors[2].get_window_position();
    return results;
}
void Enigma::set_window_positions(const std::string& new_positions)
{
    rotors[0].set_window_position(new_positions[0]);
    rotors[1].set_window_position(new_positions[1]);
    rotors[2].set_window_position(new_positions[2]);
}

void Enigma::step_rotors()
{
    if (rotors[1].in_turnover_position())
    {
        rotors[0].step_rotor();
        rotors[1].step_rotor();
        rotors[2].step_rotor();
    }
    else if (rotors[2].in_turnover_position())
    {
        rotors[1].step_rotor();
        rotors[2].step_rotor();
    }
    else
    {
        rotors[2].step_rotor();
    }
}
std::string Enigma::process_message(const std::string& message)
{
    std::string result;
    for (auto c : message)
    {
        char result_c = push_button(c);
        if (result_c == 'X')
        {
            result_c = ' ';
        }
        result += result_c;
    }
    return result;
}

char Enigma::push_button(char c)
{
    step_rotors();
    return translate_char(c);
}

char Enigma::translate_char(char c) const
{
    char forward_c = plugboard.translate_char(c);
    forward_c = rotors[2].translate_forward(forward_c);
    forward_c = rotors[1].translate_forward(forward_c);
    forward_c = rotors[0].translate_forward(forward_c);
    char reverse_c = reflector.translate_forward(forward_c);
    reverse_c = rotors[0].translate_reverse(reverse_c);
    reverse_c = rotors[1].translate_reverse(reverse_c);
    reverse_c = rotors[2].translate_reverse(reverse_c);
    char final_char = plugboard.translate_char(reverse_c);
    return final_char;
}

File: python/enigma.py
from rotor import Rotor
from rotor import Wheel
from rotor import UKW_B_WIRING
from rotor import rotor_descriptions
from messages import messages
from plugboard import Plugboard

class Enigma:
    def __init__(self, codebook_rotors, codebook_ring_settings, codebook_plugboard_settings):
        rotor_names = codebook_rotors.split(' ')
        ring_settings = codebook_ring_settings.split(' ')
        if len(rotor_names) != 3:
            raise Exception('%s did not yeild 3 rotors' % codebook_rotors)
        if len(ring_settings) != 3:
            raise Exception('%s did not yeild 3 ring settings' % ring_settings)
        self.rotors = []
        for i in range(3):
            r = rotor_names[i]
            ring_setting = int(ring_settings[i])
            wiring = rotor_descriptions[r]['wiring']
            turnovers = rotor_descriptions[r]['turnovers']
            rotor = Rotor(wiring, turnovers, ring_setting)
            self.rotors.append(rotor)
        self.plugboard = Plugboard(codebook_plugboard_settings)
        self.reflector = Wheel(UKW_B_WIRING)
    def set_window_positions(self, new_positions):
        for i in range(3):
            self.rotors[i].set_window_position(new_positions[i])
    def get_window_positions(self):
            results = ''
            for rotor in self.rotors:
                results+=(rotor.get_window_position())
            return results
    def step(self):
        do_step = [False, False, True]
        if self.rotors[1].in_turnover_position():
            do_step[0] = True
            do_step[1] = True
        if self.rotors[2].in_turnover_position():
            do_step[1] = True
        for i in range(3):
            if do_step[i]:
                self.rotors[i].step()
    def translate_char(self, c):
        forward = self.plugboard.translate(c)
        for i in [2,1,0]:
            forward = self.rotors[i].translate_forward(forward)
        reverse = self.reflector.translate_forward(forward)
        for i in [0, 1, 2]:
            reverse = self.rotors[i].translate_reverse(reverse)
        final = self.plugboard.translate(reverse)
        return final
    def press_key(self, c):
        self.step()
        return self.translate_char(c)
    def process_string(self, s):
        results = ''
        for c in s:
            results += self.press_key(c)
        return results
def self_test():
    enigma = Enigma('III II I', '1 1 1', '')
    enigma.set_window_positions('ADO')
    print enigma.get_window_positions()
    for i in range(6):
        enigma.step()
        print enigma.get_window_positions()
    for message in messages:
        enigma = Enigma(message.rotors, message.ring_settings(), message.plugboard)
        enigma.set_window_positions(message.window_positions)
        print 'Decoding ->%s<-' % message.ciphertext()
        raw_decrypt = enigma.process_string(message.ciphertext())
        print 'Raw_decrypt: %s' % raw_decrypt
        decrypt = raw_decrypt.replace('X', ' ')
        print 'Decrypt: %s' % decrypt
        print 'Final Rotor Positions: %s' % enigma.get_window_positions()
if __name__ == '__main__':
    self_test()    
